apps/onnx/model.cpp

```diff
@@ -2,483 +2,573 @@
+  for (int i = 0; i < num_iters; ++i) {
+    cache_evictor.flush_caches();
+  }
+  clock_gettime(CLOCK_REALTIME, &end);
+  double input_gen_time =
+      (end.tv_sec - start.tv_sec) * 1e9 + end.tv_nsec - start.tv_nsec;
+
+  total_runtime -= input_gen_time;
+
+  // Return the average runtime. TODO: filter the outliers if any.
+  return total_runtime / num_iters;
+}
 
-    // Now benchmark by computing the value of the outputs num_iter times
-    struct timespec start;
-    struct timespec end;
-    clock_gettime(CLOCK_REALTIME, &start);
-    for (int i = 0; i < num_iters; ++i) {
-        pipeline.rep->realize(real, tgt);
-    }
-    clock_gettime(CLOCK_REALTIME, &end);
-
-    double total_runtime =
-        (end.tv_sec - start.tv_sec) * 1e9 + end.tv_nsec - start.tv_nsec;
-
-    // Return the average runtime. TODO: filter the outliers if any.
-    return total_runtime / num_iters;
+void compile(
+    const HalideModel& pipeline,
+    const std::string& func_name,
+    const std::string& lib_name) {
+  std::vector<Halide::Argument> inputs;
+  for (const std::string& input_name : pipeline.input_names) {
+    inputs.push_back(pipeline.model->inputs.at(input_name));
+  }
+  Halide::Target tgt = Halide::get_host_target();
+  // tgt.set_feature(Halide::Target::Debug, true);
+  // tgt.set_feature(Halide::Target::NoBoundsQuery, true);
+  // tgt.set_feature(Halide::Target::TracePipeline, true);
+  // tgt.set_feature(Halide::Target::TraceRealizations, true);
+  // pipeline.rep->compile_to_lowered_stmt(std::string("/tmp/")+lib_name+".stmt",
+  // inputs, Halide::Text, tgt);
+  pipeline.rep->compile_to_file(
+      std::string("/tmp/") + lib_name, inputs, func_name, tgt);
+  pipeline.rep->compile_to_static_library(
+      std::string("/tmp/") + lib_name, inputs, func_name, tgt);
+  pipeline.rep->compile_to_c(
+      std::string("/tmp/") + lib_name + ".cpp", inputs, func_name, tgt);
+  pipeline.rep->compile_to_header(
+      std::string("/tmp/") + lib_name + ".h", inputs, func_name, tgt);
 }
 
-void print_loop_nest(const HalideModel &pipeline) {
-    pipeline.rep->print_loop_nest();
+void print_loop_nest(const HalideModel& pipeline) {
+  pipeline.rep->print_loop_nest();
 }
 
-void print_lowered_statement(const HalideModel &pipeline) {
-    std::string tmp_file = std::tmpnam(nullptr);
-    pipeline.rep->compile_to_lowered_stmt(tmp_file, pipeline.rep->infer_arguments());
-    std::ifstream is(tmp_file);
-    std::string line;
-    while (std::getline(is, line)) {
-        std::cout << line << std::endl;
-    }
-    std::remove(tmp_file.c_str());
+void print_lowered_statement(const HalideModel& pipeline) {
+  std::string tmp_file = std::tmpnam(nullptr);
+  pipeline.rep->compile_to_lowered_stmt(
+      tmp_file, pipeline.rep->infer_arguments());
+  std::ifstream is(tmp_file);
+  std::string line;
+  while (std::getline(is, line)) {
+    std::cout << line << std::endl;
+  }
+  std::remove(tmp_file.c_str());
 }
 
-}  // namespace
+} // namespace
```

Comment:
Here and elsewhere: for a trailing `// comment` put exactly two spaces before the //